modules/rules_rust/0.34.1/MODULE.bazel

```diff
@@ -0,0 +1,144 @@
+    "cui__tempfile-3.8.1",
+    "cui__tera-1.19.1",
+    "cui__textwrap-0.16.0",
+    "cui__toml-0.8.6",
+    "cui__tracing-0.1.40",
+    "cui__tracing-subscriber-0.3.17",
+    "generated_inputs_in_external_repo",
+    "libc",
+    "llvm-raw",
+    "rrra__anyhow-1.0.71",
+    "rrra__clap-4.3.11",
+    "rrra__env_logger-0.10.0",
+    "rrra__itertools-0.11.0",
+    "rrra__log-0.4.19",
+    "rrra__serde-1.0.171",
+    "rrra__serde_json-1.0.102",
+    "rules_rust_bindgen__bindgen-0.65.1",
+    "rules_rust_bindgen__bindgen-cli-0.65.1",
+    "rules_rust_bindgen__clang-sys-1.6.1",
+    "rules_rust_bindgen__clap-4.3.3",
+    "rules_rust_bindgen__clap_complete-4.3.1",
+    "rules_rust_bindgen__env_logger-0.10.0",
+    "rules_rust_prost",
+    "rules_rust_prost__h2-0.3.19",
+    "rules_rust_prost__heck",
+    "rules_rust_prost__prost-0.11.9",
+    "rules_rust_prost__prost-types-0.11.9",
+    "rules_rust_prost__protoc-gen-prost-0.2.2",
+    "rules_rust_prost__protoc-gen-tonic-0.2.2",
+    "rules_rust_prost__tokio-1.28.2",
+    "rules_rust_prost__tokio-stream-0.1.14",
+    "rules_rust_prost__tonic-0.9.2",
+    "rules_rust_test_load_arbitrary_tool",
+    "rules_rust_tinyjson",
+    "rules_rust_toolchain_test_target_json",
+    "rules_rust_util_import__aho-corasick-0.7.15",
+    "rules_rust_util_import__lazy_static-1.4.0",
+    "rules_rust_util_import__proc-macro2-1.0.33",
+    "rules_rust_util_import__quickcheck-1.0.3",
+    "rules_rust_util_import__quote-1.0.10",
+    "rules_rust_util_import__syn-1.0.82",
+    "rules_rust_wasm_bindgen__anyhow-1.0.71",
+    "rules_rust_wasm_bindgen__assert_cmd-1.0.8",
+    "rules_rust_wasm_bindgen__diff-0.1.13",
+    "rules_rust_wasm_bindgen__docopt-1.1.1",
+    "rules_rust_wasm_bindgen__env_logger-0.8.4",
+    "rules_rust_wasm_bindgen__log-0.4.19",
+    "rules_rust_wasm_bindgen__predicates-1.0.8",
+    "rules_rust_wasm_bindgen__rayon-1.7.0",
+    "rules_rust_wasm_bindgen__rouille-3.6.2",
+    "rules_rust_wasm_bindgen__serde-1.0.171",
+    "rules_rust_wasm_bindgen__serde_derive-1.0.171",
+    "rules_rust_wasm_bindgen__serde_json-1.0.102",
+    "rules_rust_wasm_bindgen__tempfile-3.6.0",
+    "rules_rust_wasm_bindgen__ureq-2.8.0",
+    "rules_rust_wasm_bindgen__walrus-0.19.0",
+    "rules_rust_wasm_bindgen__wasm-bindgen-0.2.88",
+    "rules_rust_wasm_bindgen__wasm-bindgen-cli-support-0.2.88",
+    "rules_rust_wasm_bindgen__wasm-bindgen-shared-0.2.88",
+    "rules_rust_wasm_bindgen__wasmparser-0.102.0",
+    "rules_rust_wasm_bindgen__wasmprinter-0.2.60",
+    "rules_rust_wasm_bindgen_cli",
+)
+
+rust = use_extension("//rust:extensions.bzl", "rust")
+
+# Allow us to run, for example, "bazel build //tools/runfiles" with bzlmod.
+# Register it as a dev dependency so that we don't force this toolchain on
+# downstream users.
+rust.toolchain(edition = "2021")
+use_repo(rust, "rust_toolchains")
+
+register_toolchains(
+    "@rust_toolchains//:all",
+    dev_dependency = True,
+)
+
+use_repo(rust, "rust_host_tools")
```

Comment:
Do you really want to declare the `toolchain` tag as a non-dev dependency, but then only register the toolchain as a dev dependency? Based on the comment, I would have expected the `rust` extension usage to also come with `dev_dependency = True`.

What is the story around default toolchain registrations for rules_rust? If a ruleset needs to compile a rust tool from source, can it rely on some "reasonable" Rust toolchain being registered even if the root module itself doesn't use rules_rust?

Reply:
Ah yeah, it should probably either be use_extension as a dev dependency or register toolchains without a dev dependency. My intention was to ensure that it was explicit and that you didn't accidentally use the default toolchain when you'd intended to use a custom one, so I'd intended for use_extension to be a dev dependency.

However, I'd like to hear your opinion on best practices. Is one option preferred over the other?

Reply:
As Xudong likes to put it, all of us are just "winging it" when it comes to Bzlmod - it's too early proven best practices. What I found to work well so far is to have rulesets with relatively canonical SDKs (mostly not C++) promise to always register some reasonably recent version of a hermetic SDK so that transitive deps of a project can use `rust_binary` without placing the toolchain resolution burden on the root module. That does seem to work for Go, but whether it makes sense for Rust depends a lot on the ecosystem.

Reply:
I believe this is fixed by https://github.com/bazelbuild/rules_rust/pull/2351.

Reply:
Does this still allow the root module to override the toolchain used?

Reply:
Yes. According to the bazel docs, the toolchain resolution is:

1. Toolchains registered using --extra_toolchains are added first. (Within these, the last toolchain has highest priority.)
2. Toolchains registered using register_toolchains in the transitive external dependency graph, in the following order: (Within these, the first mentioned toolchain has highest priority.)
3. Toolchains registered by the root module (as in, the MODULE.bazel at the workspace root);
4. Toolchains registered in the user's WORKSPACE file, including in any macros invoked from there;
5. Toolchains registered by non-root modules (as in, dependencies specified by the root module, and their dependencies, and so forth);
6. Toolchains registered in the "WORKSPACE suffix"; this is only used by certain native rules bundled with the Bazel installation.

All you need to do is to call register_toolchains from the root module, and you should be able to override the toolchains.